File: padroesEstruturais/padrao-de-projeto-composite/php/src/Componentes/Componente.php
<?php

namespace Composite\Php\Componentes;

use Composite\Php\Pecas\Peca;
use Ds\Vector;

abstract class Componente
{
    public function __construct(
        private Vector $pecas = new Vector(),
        private float $valor = 0.0,
        private string $descricao = '',
    ) {
    }

    public function setPecas(Peca $peca): Componente
    {
        $this->pecas->push($peca);
        return $this;
    }

    public function getPecas(): string
    {
        $saida = '';
        /** @var Peca $peca */
        foreach($this->pecas as $peca) {
            $saida .= $peca->getDescricao();
        }

        return $saida;
    }

    public function setDescricao(string $descricao): Componente
    {
        $this->descricao = $descricao;
        return $this;
    }

    public function getDescricao(): string
    {
        $saida = $this->descricao . " / valor: " . $this->getValor() . "\nPeças: \n";
        /** @var Peca $peca */
        foreach($this->pecas as $peca) {
            $saida .= $peca->getDescricao() . "\n";
        }
        return $saida;
    }

    public function setValor(float $valor): Componente
    {
        $this->valor = $valor;
        return $this;
    }

    public function getValor(): float
    {
        return $this->valor;
    }
}

File: padroesEstruturais/padrao-de-projeto-composite/java/src/main/java/composite/Componentes/Componente.java
package composite.Componentes;

import composite.Pecas.Peca;

import java.util.ArrayList;
import java.util.List;

public abstract class Componente {

    private List<Peca> pecas = new ArrayList<Peca>();
    private String descricao;
    private Float valor;

    public Componente setPecas(Peca peca) {
        this.pecas.add(peca);
        return this;
    }

    public String getPecas() {
        StringBuilder saida = new StringBuilder();
        for(Peca pecas : this.pecas) {
            saida.append(pecas.getDescricao()).append(" / ");
        }

        return saida.toString();
    }

    public Componente setDescricao(String descricao) {
        this.descricao = descricao;
        return this;
    }

    public String getDescricao() {
        StringBuilder saida = new StringBuilder(this.descricao + " / valor: " + this.getValor() + "\nPeças: \n");
        for(Peca peca : this.pecas) {
            saida.append(peca.getDescricao()).append("\n");
        }

        return saida.toString();
    }

    public Componente setValor(Float valor) {
        this.valor = valor;
        return this;
    }

    public Float getValor() {
        return this.valor;
    }
}
